File: 16236번: 아기 상어/아기 상어.cpp
#include <iostream>
#include <vector>
#include <queue>

#define point pair<int, int> 
#define vvi vector<vector<int>>
#define MAX 100000000

using namespace std;

int N;
vvi graph;
point shark;
const point directions[] = {{0,1}, {0,-1},{1,0},{-1,0}};

int bfs(vvi&, vvi&, int);


int main() {
    cin >> N;
    graph = vvi(N, vector<int>(N));
    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            cin >> graph[i][j];
            if (graph[i][j] == 9) {
                shark = {i, j};
                graph[i][j] = 0;
            }
        }
    }

    vvi visited(N, vector<int>(N));
    int seconds = 0, sec = 0;
    int num_of_eating = 0, shark_value = 2;
    while (true) {
        for (int i=0; i<N; i++) {
            fill(visited[i].begin(), visited[i].end(), MAX);
        }
        
        if (sec = bfs(visited, graph, shark_value)) {
            num_of_eating += 1;
            seconds += sec;
            if (num_of_eating == shark_value) {
                ++shark_value;
                num_of_eating = 0;
            }
        }
        else {
            cout << seconds << endl;
            break;
        }
    }

    return 0;
}

int bfs(vvi &visited, vvi &graph, int shark_value) {
    priority_queue<pair<int, point>> queue;
    queue.push({0, shark});
    visited[shark.first][shark.second] = 0;
    int min_dist = MAX;

    while (!queue.empty()) {
        auto [dist, cur] = queue.top();
        queue.pop();

        dist = -1 * dist;
        
        if (dist > min_dist) continue;
        
        int v = graph[cur.first][cur.second];
        if (0<v && v<shark_value) {
            min_dist = min(min_dist, dist);
        }

        int ndist = dist + 1;

        for (auto di : directions) {
            int nr = cur.first + di.first;
            int nc = cur.second + di.second;
            if (0<=nr && 0<=nc && nr<N && nc<N && graph[nr][nc]<=shark_value) {
                if (visited[nr][nc] > ndist) {
                    visited[nr][nc] = ndist;
                    queue.push({-ndist, {nr, nc}});
                }
            }
        }
    }

    for (int i=0; i<N; i++) {
        for (int j=0; j<N; j++) {
            if (min_dist<MAX && min_dist==visited[i][j]) { 
                if (graph[i][j] < shark_value && 0< graph[i][j]) {
                    graph[i][j] = 0;
                    shark = {i, j};
                    return min_dist;
                }
            }
        }
    }

    return 0;
}

File: 16236번: 아기 상어/아기 상어.py
#  **************************************************************************  #
#                                                                              #
#                                                       :::    :::    :::      #
#    Problem Number: 16236                             :+:    :+:      :+:     #
#                                                     +:+    +:+        +:+    #
#    By: ydh9516 <boj.kr/u/ydh9516>                  +#+    +#+          +#+   #
#                                                   +#+      +#+        +#+    #
#    https://boj.kr/16236                          #+#        #+#      #+#     #
#    Solved: 2025/04/09 14:06:14 by ydh9516       ###          ###   ##.kr     #
#                                                                              #
#  **************************************************************************  #
import sys
from heapq import heappop, heappush
input = sys.stdin.readline

N = int(input())
graph = [list(map(int, input().split())) for _ in range(N)]

directions = [(-1,0),(0,-1),(0,1),(1,0)]
shark = [(i, j) for i in range(N) for j in range(N) if graph[i][j]==9][0]

MAX=100_000_000
def bfs(shark, shark_value):
    global directions, graph, N
    visited = [[MAX]*N for _ in range(N)]
    min_val = MAX
    que = []
    que.append((0, shark))
    visited[shark[0]][shark[1]] = 0
    while que:
        sec, cur = heappop(que)
        sec *= -1
        nsec = sec + 1
        
        if sec > min_val: 
            continue
        
        if 0 < graph[cur[0]][cur[1]] < shark_value:
            min_val = min(min_val, sec)
                
        for di in directions:
            nr = cur[0] + di[0]
            nc = cur[1] + di[1]
            if 0<=nr<N and 0<=nc<N and shark_value >= graph[nr][nc]:
                if visited[nr][nc] > nsec:
                    visited[nr][nc] = nsec
                    heappush(que, (-nsec,(nr,nc)))

    for i in range(N): 
        for j in range(N):
            if visited[i][j] == min_val < MAX and 0 < graph[i][j] < shark_value:
                return (i, j), min_val
    
    return None, -1

seconds = 0
graph[shark[0]][shark[1]] = 0
shark_value = 2
num_of_eating = 0
i = 0
while True:
    i+=1
    new_feed, sec = bfs(shark, shark_value)
    if new_feed is None: break

    # print(i, "new_feed", new_feed ,"cur shark:", shark_value)
    graph[new_feed[0]][new_feed[1]] = 0
    shark = new_feed
    num_of_eating += 1
    if num_of_eating == shark_value:
        shark_value += 1
        num_of_eating = 0

    seconds += sec

print(seconds)


